[ES|QL] Technical preview in Discover and Dashboards

stratoula opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/137810
Part of https://github.com/elastic/kibana/issues/163248

This PR is the **first iteration** of the new ES|QL language in kibana. The majority of the functionality is based on the existing functionality for SQL (which has already been merged). This builds on top of it to enable the functionality for ES|QL. We decided to remove SQL for now so this PR enables ES|QL and hides SQL. We are not removing the code for now.

### Important notes:
- This PR contains basic autocomplete functionality for the ES|QL commands. We want to improve it in follow up PRs
- The majority of the tests for SQL were moved to work with ES|QL instead
- The search strategy is a very simple endpoint for now as we don't have async search nor pagination
- Now that we remove SQL, the ui for ES|QL selection has changed
- The documentation for ESQL has been handled by the docs team so it is already reviewed and in sync with the official documentation
- ES|QL is disabled in serverless projects for now

### Changes from SQL:
- The Discover histogram now is being created with ES|QL (using the date_trunc function). This gives it the ability to be saved on a dashboard and also to be edited inline.
- ES|QL sometimes returns some warnings (on the search headers). For example when we are trying to date parse a string that doesnt contain a valid date). These warnings are also reported on the UI. There is a bug in ES and the warning doesnt come always https://github.com/elastic/elasticsearch-internal/issues/1465)
- We have 2 types of commands:
  - Transformational commands (keep, stats) -> when they exist in the query Discover should render the Table view (selected columns)
  - Non transformational commands (all the others) -> in that case Discover renders the Document view
  - ESQL switch on advanced settings is now on by default


**Discover view with non transformational commands**
<img width="1678" alt="image" src="https://github.com/elastic/kibana/assets/17003240/abe100e1-01e9-4fe0-9b89-6d8bdf6443fc">


**Discover view with transformational commands**
<img width="1679" alt="image" src="https://github.com/elastic/kibana/assets/17003240/e46af422-daeb-4be5-88cf-522211674ff5">


### Missing
- ESQL autocomplete is not perfect, we are going to work on it on a future PR. Specifically:
  - There is a bug in autocomplete, sometimes writing a query overwrites the existing one, will deal with it in a future PR
  - Further improvements

### Checklist

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

alexwizp commented:
@elasticmachine merge upstream

stratoula commented:
Sorry for pinging you here, it was by mistake. There is a long run since this PR goes public!

afharo reviewed (approved):
Telemetry changes LGTM

stratoula commented:
@elasticmachine merge upstream

stratoula commented on x-pack/test/search_sessions_integration/tests/apps/discover/async_search.ts:
```diff
@@ -131,6 +131,19 @@ export default function ({ getPageObjects, getService }: FtrProviderContext) {
 
       expect(searchesCountBeforeRestore).to.be(searchesCountAfterRestore); // no new searches started during restore
     });
+
```
ℹ️ I had removed it in a previous PR (when we were using the same histogram from the dataview mode). Now that the histogram is constructed from an ES|QL query, I am adding this back

stratoula commented on x-pack/test/search_sessions_integration/tests/apps/discover/async_search.ts:
```diff
@@ -131,6 +131,19 @@ export default function ({ getPageObjects, getService }: FtrProviderContext) {
     });
+
+    it('should should clean the search session when navigating to SQL mode, and reinitialize when navigating back', async () => {
+      await PageObjects.common.navigateToApp('discover');
```
ℹ️ I had removed it in a previous PR (when we were using the same histogram from the dataview mode). Now that the histogram is constructed from an ES|QL query, I am adding this back

stratoula commented on src/plugins/unified_search/public/dataview_picker/explore_matching_button.tsx:
```diff
@@ -35,7 +35,6 @@ export const ExploreMatchingButton = ({
         alignItems="center"
         gutterSize="none"
         justifyContent="spaceBetween"
-        data-test-subj="select-text-based-language-panel"
```
ℹ️ Irrelevant with the ESQL enablement, I just removed the data subj because it is not used and is also wrong

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

stratoula commented on test/functional/apps/discover/group3/_request_counts.ts:
```diff
@@ -144,7 +146,7 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
         await waitForLoadingToFinish();
         // creating the saved search
-        await expectSearches(type, 2, async () => {
+        await expectSearches(type, savedSearchesRequests ?? 2, async () => {
```
ℹ️ I dont understand why it runs 4 times on a saved search, I am looking into it from yesterday but I am def something missing. I would love the help from the https://github.com/orgs/elastic/teams/kibana-data-discovery team as they know this code better than me. We can do it on a follow up PR as I would love to merge the feature branch as soon as possible as multiple PRs depend on it.

kertal replied:
had a quick look at it, without having a solution. Agreed that this can be done in a follow up PR by @elastic/kibana-data-discovery

kertal replied:
ideally we could  add  a TODO comment with a linked issue

stratoula replied:
I will do that :) Thanx for checking!

stratoula replied:
Done!

ninoslavmiskovic commented on docs/management/advanced-options.asciidoc:
Very very small "nit" - Is it possible to have the "|" or is that an illegal character? It is displayed in the UI like `discover:enableESQL` and perhaps it could be `discover:enableES|QL
 `

stratoula replied:
It is an illegal character, this is why I preferred to be like that. Otherwise it will cause a lot of problems

kertal commented on docs/management/advanced-options.asciidoc:
```diff
@@ -288,8 +288,8 @@ in the current data view is used.
-[[discover:enableSql]]`discover:enableSql`::
-experimental[] Allows SQL queries for search.
+[[discover:enableESQL]]`discover:enableESQL`::
+experimental[] Allows ESQL queries for search.
```
@ninoslavmiskovic @stratoula I think it's less illegal here, right? 🚓 
```suggestion
experimental[] Allows ES|QL queries for search.
```

davismcphee commented on packages/kbn-text-based-editor/src/helpers.ts:
```diff
@@ -101,6 +138,21 @@ export const getDocumentationSections = async (language: string) => {
+      await import('./esql_documentation_sections');
+    groups.push({
+      label: i18n.translate('textBasedEditor.query.textBasedLanguagesEditor.esql', {
+        defaultMessage: 'ESQL',
```
```suggestion
        defaultMessage: 'ES|QL',
```

davismcphee commented on src/plugins/unified_search/public/dataview_picker/change_dataview.tsx:
```diff
@@ -335,42 +331,24 @@ export function ChangeDataView({
+            data-test-subj="select-text-based-language-panel"
+          >
+            {i18n.translate('unifiedSearch.query.queryBar.textBasedLanguagesTryLabel', {
+              defaultMessage: 'Try ES | QL',
```
```suggestion
              defaultMessage: 'Try ES|QL',
```
Should there be spaces here?

stratoula replied:
No without spaces is the correct here, you are right 👍

davismcphee commented on x-pack/plugins/lens/public/datasources/text_based/utils.test.ts:
```diff
@@ -66,6 +66,22 @@ describe('Text based languages utils', () => {
+
+    it('should return the index pattern for es|ql query', () => {
+      const indexPattern = getIndexPatternFromTextBasedQuery({
+        sql: 'from foo | keep bytes, memory ',
```
```suggestion
        esql: 'from foo | keep bytes, memory ',
```
Should this be `esql`?

davismcphee reviewed (commented):
Did an initial code review and it's looking really good overall! I didn't see anything that jumped out to me as an issue, and mostly just left some minor suggestions and nits for now. Will followup soon with feedback from my local testing.

stratoula commented:
@davismcphee thanx a ton for the thorough code review! I have addressed everything here https://github.com/elastic/kibana/pull/146971/commits/6b75898ce1d529658b2753725fe75b6ab57539f6

dej611 commented on packages/kbn-text-based-editor/src/text_based_languages_editor.styles.ts:
```diff
@@ -40,7 +42,11 @@ export const textBasedLanguagedEditorStyles = (
     resizableContainer: {
       display: 'flex',
-      width: isCodeEditorExpanded ? '100%' : 'calc(100% - 80px)',
+      width: isCodeEditorExpanded
+        ? '100%'
+        : hasReference
+        ? 'calc(100% - 80px)'
+        : 'calc(100% - 40px)',
```
nit:
```suggestion
      width: isCodeEditorExpanded
        ? '100%'
        : `calc(100% - ${hasReference ? 80 : 40}px)`,
```

dej611 commented on src/plugins/discover/public/application/main/components/layout/discover_documents.tsx:
```diff
@@ -185,10 +185,11 @@ function DiscoverDocumentsComponent({
     () =>
-      !isTextBasedQuery &&
+      // for ES|QL we want to show the time column only when is on Document view
+      (!isTextBasedQuery || (isTextBasedQuery && !columns?.length)) &&
```
Isn't implicit the positive `isTextBasedQuery` on the right handside?
```suggestion
      (!isTextBasedQuery || !columns?.length) &&
```

dej611 reviewed (approved):
Code review 👍 

Left few nitpick comments but not a blocker 🎉

stratoula commented:
Thanx @dej611, applied your comments here https://github.com/elastic/kibana/pull/146971/commits/d81042ef416c5cd1d5422c22c475f3d2f986da50

davismcphee reviewed (approved):
Fantastic work on this, it's so exiting to finally see ES|QL in Discover! I tested locally pretty extensively today and for the most part didn't encounter any broken functionality or major bugs. I did notice two main issues though:

You mentioned this in the summary regarding autocomplete issues:
>There is a bug in autocomplete, sometimes writing a query overwrites the existing one, will deal with it in a future PR

I ran into this quite often in my testing where after autocompleting, the ES|QL editor would be completely bugged out and overwrite characters, not accept characters, lose syntax highlighting, or even get out of sync. It sounds like this is a known issue with plans to improve it, so I won't block the merge over it, but as a user it was quite frustrating, so hopefully we can get it sorted quickly.

I also noticed a lot of scenarios where duplicate requests are being fired in ES|QL mode, sometimes up to 3 extra fetches at once. Again since it's technical preview, I'll leave it up to you to decide if this is something we're ok living with for now (data views seem unaffected which is the main thing to me), but I would definitely suggest we prioritize addressing this soon.

I see there's already an issue for some of the extra fetches in #165192, but I noticed a number of other scenarios in my testing too:
- 1 extra fetch on chart hide
- 1 extra fetch on page reload when a chart is visible
- 1 extra fetch when changing from an index where the vis is hidden to one where it's shown
- 3 extra fetches when navigating back from Dashboard to Discover after adding an ES|QL Discover vis
- 3 extra fetches when saving or opening a saved search

Maybe we can add these cases to #165192 and expand it into an issue for addressing the extra fetch issue more broadly? I may be able to help with some of this once I can find some time for it.

One thing I noticed that I think existed before this PR but would be good to make a note to fix is the disabled time picker tooltip uses `esql` instead of `ES|QL`:
<img width="277" alt="popover" src="https://github.com/elastic/kibana/assets/25592674/b647c979-eafd-4446-ad95-a6ffe9f714e3">

Another thing that would be good followup work for another PR is adding some details to the ES|QL request for the inspector. I saw you mentioned the ES|QL search implementation is pretty basic for now, maybe something to consider when evolving this:
<img width="605" alt="request" src="https://github.com/elastic/kibana/assets/25592674/79c226b5-302b-44ce-bf23-55fe244ef455">

Otherwise I created a very small PR with minor cleanup of the remaining references to `SQL` I found in Discover and Unified Histogram (no functionality changes): #165426. Feel free to merge to this PR if you want, or I can point it at `main` instead once this is merged.

And lastly I left a couple of suggestions for places where additional unit tests would be ideal. I get there's some urgency on this PR so again I won't block on it, but it would be great to consider improving the testing in those areas when we can.

My only other feedback is that it would be great to get a quick review from @lukasolson on the search changes if possible since he's most familiar with the area on our team.

Thanks again for all the effort on this @stratoula! This really is an amazing feature and milestone for Kibana and Discover 🙌

stratoula commented on src/plugins/unified_histogram/public/layout/hooks/use_lens_suggestions.ts:
```diff
@@ -50,6 +62,63 @@ export const useLensSuggestions = ({
+        textBasedColumns: [
+          {
+            id: `${dataView.timeFieldName} every ${interval}`,
+            name: `${dataView.timeFieldName} every ${interval}`,
```
This can't be translated unfortunately. It is not used as label on the Lens suggestion api. It is the variable generated from the query. We dont support custom labels at the moment in text based mode. It is def one of the enhancements we could do.

stratoula commented on src/plugins/unified_histogram/public/layout/hooks/use_lens_suggestions.ts:
```diff
@@ -50,6 +62,63 @@ export const useLensSuggestions = ({
+          },
+          {
+            id: 'rows',
+            name: 'rows',
```
Same here

stratoula commented:
@davismcphee thanx a lot for the review

> I also noticed a lot of scenarios where duplicate requests are being fired in ES|QL mode, sometimes up to 3 extra fetches at once. 
I updated the issue. We should def look into this but I will need some help so we can look into it together when you have some time

> One thing I noticed that I think existed before this PR but would be good to make a note to fix is the disabled time picker tooltip uses esql instead of ES|QL:

Updated!

> Another thing that would be good followup work for another PR is adding some details to the ES|QL request for the inspector. I saw you mentioned the ES|QL search implementation is pretty basic for now, maybe something to consider when evolving this:

The ES team wants to work on async search of the esql api. At this point we can discuss together how we want to implement it with some help from @lukasolson as he knows this area better and we can revisit the inspector.

> Otherwise I created a very small PR with minor cleanup of the remaining references to SQL I found in Discover and Unified Histogram (no functionality changes)

Thanx a ton for this! I will merge when CI is green

> And lastly I left a couple of suggestions for places where additional unit tests would be ideal. I get there's some urgency on this PR so again I won't block on it, but it would be great to consider improving the testing in those areas when we can.

I will create an issue and we can about it

stratoula commented on src/plugins/unified_histogram/public/layout/hooks/use_lens_suggestions.ts:
Created a follow up issue https://github.com/elastic/kibana/issues/165431

stratoula commented on src/plugins/discover/public/application/main/hooks/use_text_based_query_language.ts:
Created a follow up issue https://github.com/elastic/kibana/issues/165431

stratoula commented:
@elasticmachine merge upstream

stratoula commented:
I am merging it as the serverless obs tests failures are irrelevant with this PR changes
